crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -721,54 +749,24 @@ fn deserialize_modules<S: StorageView, Mode: ExecutionMode>(
                 b,
                 context.protocol_config.move_binary_format_version(),
             )
-            .map_err(|e| e.finish(move_binary_format::errors::Location::Undefined))
+            .map_err(|e| e.finish(Location::Undefined))
         })
-        .collect::<move_binary_format::errors::VMResult<Vec<CompiledModule>>>()
+        .collect::<VMResult<Vec<CompiledModule>>>()
         .map_err(|e| context.convert_vm_error(e))?;
 
     assert_invariant!(
         !modules.is_empty(),
         "input checker ensures package is not empty"
     );
-    Ok(modules)
-}
-
-/// - Deserializes the modules
-/// - Publishes them into the VM, which invokes the Move verifier
-/// - Run the Sui Verifier
-fn publish_and_verify_new_modules<S: StorageView, Mode: ExecutionMode>(
-    context: &mut ExecutionContext<S>,
-    module_bytes: &[Vec<u8>],
-) -> Result<Vec<CompiledModule>, ExecutionError> {
-    let mut modules = deserialize_modules::<_, Mode>(context, module_bytes)?;
 
-    // It should be fine that this does not go through ExecutionContext::fresh_id since the Move
-    // runtime does not to know about new packages created, since Move objects and Move packages
-    // cannot interact
-    let package_id = if Mode::packages_are_predefined() {
-        // do not calculate package id for genesis modules
-        (*modules[0].self_id().address()).into()
-    } else {
-        generate_package_id(&mut modules, context.tx_context)?
-    };
-    publish_and_verify_modules(context, package_id, modules)
-}
-
-fn publish_and_verify_upgraded_modules<S: StorageView, Mode: ExecutionMode>(
-    context: &mut ExecutionContext<S>,
-    module_bytes: &[Vec<u8>],
-    package_id: ObjectID,
-) -> Result<Vec<CompiledModule>, ExecutionError> {
-    let mut modules = deserialize_modules::<_, Mode>(context, module_bytes)?;
-    substitute_package_id(&mut modules, package_id)?;
-    publish_and_verify_modules(context, package_id, modules)
+    Ok(modules)
 }
 
 fn publish_and_verify_modules<S: StorageView>(
     context: &mut ExecutionContext<S>,
     package_id: ObjectID,
-    modules: Vec<CompiledModule>,
-) -> Result<Vec<CompiledModule>, ExecutionError> {
+    modules: &[CompiledModule],
+) -> Result<(), ExecutionError> {
     // TODO(https://github.com/MystenLabs/sui/issues/69): avoid this redundant serialization by exposing VM API that allows us to run the linker directly on `Vec<CompiledModule>`
```

Comment:
At this point we serialize these twice - once here and once when creating a `MoveObject`. More of an observation as I am not sure if there is an easy way to eliminate this redundancy without what the code comment above suggests.